[Color Picker] custom format implementation

donlaci opened the pull request:
<!-- Enter a brief description/summary of your PR here. What does it fix/what does it change/how was it tested (even manually, if necessary)? -->
## Summary of the Pull Request
Extend the Color Picker module with the possibility of using custom color formats.  
<!-- Please review the items on the PR checklist before submitting-->
## PR Checklist

- [X] **Closes:** #8305 
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated
- [ ] **New binaries:** Added on the required places
   - [ ] [JSON for signing](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ESRPSigning_core.json) for new binaries
   - [ ] [WXS for installer](https://github.com/microsoft/PowerToys/blob/main/installer/PowerToysSetup/Product.wxs) for new binaries and localization folder
   - [ ] [YML for CI pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/ci/templates/build-powertoys-steps.yml) for new test projects
   - [ ] [YML for signed pipeline](https://github.com/microsoft/PowerToys/blob/main/.pipelines/release.yml)
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/windows-uwp/tree/docs/hub/powertoys) and link it here: #xxx

<!-- Provide a more detailed description of the PR, other things fixed or any additional comments/features here -->
## Detailed Description of the Pull Request / Additional comments
The main changes are on the settings page, where there is a new button for creating new formats. 
The created formats appear in the list of the formats among the predefined ones.
There is a possibility to edit and delete the custom formats.

<!-- Describe how you validated the behavior. Add automated tests wherever possible, but list manual validation steps taken as well -->
## Validation Steps Performed
Tested locally

![image](https://user-images.githubusercontent.com/57342539/202456904-71ad6ee8-0e62-42bf-8ef2-0a4a92b5144e.png)

![image](https://user-images.githubusercontent.com/57342539/202456480-17843044-9733-450e-b478-ce52e0e9709b.png)

donlaci committed:
[ColorPicker] Development: custom color formats, first steps

donlaci committed:
ColorPicker development of custom format handling.

donlaci committed:
Custom color format developmnet.
Added common helper class for format string
Fixed settings loading
Added numbering if default name exists   (My format (1))

donlaci committed:
Custom color format implementation.
Extended the colorPicker settings with the format string
Updated the color to string conversion

donlaci committed:
Custom color format in color picker, development.
Adding edit, delete buttons. Implement functionality
Re-arranging settings panel (newly created formats at the top)
Implementing details (valid parameters, more format elements, more types)

donlaci committed:
Minor commit

donlaci committed:
Development color picker custom formats. "Last" steps.
Replacing hard coded english strings with resources, polishing.

donlaci committed:
Adding help to the format edit dialog.

donlaci committed:
Undoing changes unwillingly commited in Host module

htcfreek commented:
Why do we need the `=` and `;` in the format definition formula? Can't we go without it? Like only `new(R=255,G=255,B=255)`.

What character is used a decimal seperator? Can we split the formula parts in the bracket with `;` to allow `,` and `.` as decimal separator?

donlaci commented on src/settings-ui/Settings.UI/Strings/en-us/Resources.resw:
```diff
@@ -1286,9 +1286,141 @@ Made with 💗 by Microsoft and the PowerToys community.</value>
+    <value>byte value (default)</value>
+  </data>
+  <data name="Help_hexL1.Text" xml:space="preserve">
+    <value>hex lowercase one digit</value>
```
Hi! You mean in the "help section"? Yes we could use  the word "hexadecimal". 
There is not much space and I wanted to present all the possibilities.

Jay-o-Way replied:
Good reason

htcfreek commented:
> > 
> Hi @htcfreek It was just an example. You can define the formula as you like.
> For example       Color ( R =%Re, G = %Gr, B = %Bl)
> or (just as an example)         %Re ;  %Gr ;  %Bl   .......   %Na
> You can have any characters in the format string, the valid %XX parameters will be replaced by the corresponding values.

But for what do `=` and especially `;` (at the end) stands for?

I think we have a complex and a simple (`color(r=3,g=2,b=5)` example.

Jay-o-Way commented:
I think the `new`...`;` is implying a C# code -like thing

Jay-o-Way commented:
Hm, one idea/question home to mind: can those variables be used in the _name_ to?

htcfreek commented:
Sure @Jay-o-Way. But do we need to borrow a user with such things if it's not necessary. I think our main goal here should be to keep it short and simple for the user.

I think the best way for our users is that they only type the name, the formula part within the brackets and maybe a "short" name. And then the brackets and everything else we need is added automatically.

crutkas commented:
Oh heck yeah!!!! 🔥🔥🔥🔥🔥🔥

SeraphimaZykova reviewed (changes requested):
Nice work! Tested, everything works well except Delete button in Settings. I can't remove the custom format after creation.

Minor nits about unnecessary directives, what is the purpose of adding them? 
Also, what's the purpose of moving color names and formats to common?

donlaci committed:
Fixing bug unable to delete a custom format after renaming it - use the colorformat object as reference (and not the name)
Modifying the default custom formula
Removing unnecessary using directives

htcfreek commented:
@donlaci 
Are the custom formats only available in editor or for clipboard too? Can we synchronize clipboard and editor formats?

htcfreek commented:
@donlaci 
1. I think we should move the add custom button to the list view header action content.
2. In the settings group  befor the editor formats is a clipboard format setting. And there you can choose the format for direct copy (without using the editor). That's wat I mean.

donlaci commented:
@htcfreek 
Thanks for the clarification, now I got it. We well consider your suggestions.

SeraphimaZykova commented on src/settings-ui/Settings.UI.Library/ColorFormatModel.cs:
```diff
@@ -14,12 +14,25 @@ public class ColorFormatModel : INotifyPropertyChanged
+
+        public ColorFormatModel()
+        {
+            Example = "Color ( R =%Re, G = %Gr, B = %Bl)";
```
```suggestion
            Example = "Color ( R = %Re, G = %Gr, B = %Bl)";
```

SeraphimaZykova reviewed (approved):
LGTM! Tested, works well now.

donlaci committed:
Udating the default user defined color format

donlaci committed:
Removing unnecessary using directive

niels9001 commented:
Awesome work @donlaci 🚀. I have a few questions looking at this screenshot:

![image](https://user-images.githubusercontent.com/9866362/203328122-1933b674-b008-4f06-823a-6a8679e6d2a5.png)

1) Does this imply that the current formats (that come with PT) cannot be removed? Wouldn't it make more sense to remove them all and ship some of these as 'presets'? This feels a bit weird now. If not, I'd suggest splitting the format listviews.
2) Wouldn't make more sense if the list shows the actual preview of the format (i.e. 'what will I see in the ColorPicker UI)? Vs. the format itself?
- There are a couple of minor styling tweaks we could do, but we can leave that to after merging this in :).

niels9001 commented:
> > 
> 1. Yes, currently the old hard coded formats are kept fixed, removal is not implemented - as it was before. We could easily add the edit, remove buttons to them as well and allow the users to modify, delete them. I'm not sure whether it adds value to the app or it is better to have allways the preset formats. They can be hidden and it looks professional to have so many presets - but it is just my oppinion. :-)

Yeah.. I think adding those buttons (and from a UX perspective, make sure formats that are custom and ship with PT are 'equal'). I can imagine users going to start ask why this one format can be deleted, and the other cannot :).

> 2. Yes, we could have the preview of the format if it makes more sense. Just let me know and I'll update the code.

I'd say so? I think if we replicate the format that you'd see in the Color Picker Editor makes most sense?

Jay-o-Way commented:
FYI
![image](https://user-images.githubusercontent.com/65828559/203350064-b263a742-1033-4ca2-9574-da8eec66a1eb.png)

donlaci committed:
ColorPicker Implementing custom color formats: adding custom formats to the default format selection (dropdown box).

jaimecbernardo committed:
Merge branch 'main' into pr22141

jaimecbernardo committed:
Fix binding of name and example

jaimecbernardo commented:
I've merged main here due to the UI refactor that just went into main. Here's hope I didn't break anything 😅

htcfreek commented:
Can we please get updated screenshot from the settings page.

donlaci commented:
Hi @htcfreek,
Your proposal, to add the custom color formats to the list of formats for direct copy, has been implemented. Here you have some screenshots. 
![image](https://user-images.githubusercontent.com/57342539/203779567-4d8fcc9f-5845-4321-880b-6d82ff4cdd03.png)
 
![image](https://user-images.githubusercontent.com/57342539/203779723-0a495ddf-f09f-4464-be31-d7b3588bf3e5.png)

Jay-o-Way commented:
Until now, the SettingsCard with "" does not actually have any setting (or controls) inside of it. I think it should be converted to text, below the group header.

niels9001 commented:
I agree with @htcfreek 's point earlier - shouldn't the 'Add new format' button be in the top card, just like we do with Image Resizer?

That would look something like this:
![image](https://user-images.githubusercontent.com/9866362/204164184-5c624a97-aaa9-4c08-ae55-c4bc8459c5d5.png)

donlaci commented:
@niels9001 @htcfreek 
 The current state is: 
![image](https://user-images.githubusercontent.com/57342539/204267217-d018de7b-96e1-4a72-a888-af32bc3e20e2.png)

htcfreek commented:
> @niels9001 @htcfreek 
>  The current state is: 
> ![image](https://user-images.githubusercontent.com/57342539/204267217-d018de7b-96e1-4a72-a888-af32bc3e20e2.png)
> 

How can we make clear that the custom formats not exclusive to the editor? Rename/reorder the settings?

Jay-o-Way commented:
> How can we make clear that the custom formats not exclusive to the editor? Rename/reorder the settings?

Good point. At this point, I would say "drop the current header"

donlaci committed:
Custom color formats, implemented steps:
vorwarts compatibility loading settings.
Fixed UI as requested (removed one settings panel, added button to the first panel)

htcfreek commented:
@donlaci
Please update the description of the formats list too. Something like "Configure the available formats". Currently it has a text reference to the editor.

donlaci committed:
Minor change in the UI: description modified

donlaci commented:
Hi,
The latest state is shown on the pictures
What changed is 
- the description of the first panel
- instead of the rules an exxample is shown as description for the available formats (also for the default format selection)
- the presets (RGB, HEX, ...) are converted into 'custom' formats. ('Custom' is a new 'normal'). Every format is editable, deletable.

![image](https://user-images.githubusercontent.com/57342539/204576108-9a526e1d-e052-4d30-8477-7e467d2ffc94.png)

![image](https://user-images.githubusercontent.com/57342539/204577529-c9d5a381-8dfb-4128-8fb0-363a142caade.png)

htcfreek commented:
@donlaci 
Can you please rename the group header from "Editor" to "Color formats".

donlaci committed:
ColorPicker Custom Color Formats develepoment.
Added conversion from old predefined formats to customizable formats.
Extended default settings (in case settings file is deleted/corrupted).
Minor fixes.

donlaci committed:
Fixing color format parameters.
Implementing 3 different Saturation calculations, 2 Hue calculations and 2 Lightness calculations (depending color format)

donlaci committed:
Color Picker: New/Edit Color format. Fixing bug when cancelling addition/edit

donlaci committed:
ColorPicker. Updating help section, available parameters

jaimecbernardo committed:
Merge branch 'main' into dev-colorPicker-customFormat

jaimecbernardo committed:
Fix spellchecker

SeraphimaZykova reviewed (approved):
Great work!

jaimecbernardo committed:
Remove the MinWidth so that scrollviewers can be drawn

jaimecbernardo reviewed (changes requested):
I've added a commit to show the scrollbars in the popup (they were hidden due to the minWidth).
This PR still needs two changes:
- handle the situation where the user deletes all color formats. Currently we're crashing.
- up the settings file version to 2.0 or something similar.
Other than that, everything looks good. Great work!

htcfreek commented:
I thought the default formats are protected against deletion.

Can we convert setting then from v1 to v2 and does it keep the old configuration?

donlaci commented:
> * I thought the default formats are protected against deletion.
> * Can we convert setting then from v1 to v2 and does it keep the old configuration?
> * Are the last screenshots up to date?

Hi @htcfreek 
Now all formats are treated in the same way, all can be deleted.
Yes, settings conversion from v1 to v2 is implemented.
The screenshots are up to date except the help section. The list of available parameters has been extended (due different calculation of the 'same' parameters in different models, like saturation in HSI, HSB and HSL) The parameter list has been extended.

htcfreek commented:
@donlaci 
The settings page still has a left over: The settings group header "Editor".

donlaci committed:
ColorPicker bugfix: Not allowing to delete the last color format.

donlaci commented:
Hi @htcfreek 
I better put here new screenshots. The label "editor" has been replaced. I hope everything is fine.
![image](https://user-images.githubusercontent.com/57342539/205329465-3f95a3e8-2d92-4b83-9fbf-04882b721de0.png)
![image](https://user-images.githubusercontent.com/57342539/205329601-c766e4d3-a90d-4ae2-9cf4-573763937a1e.png)

jaimecbernardo reviewed (approved):
LGTM! Great work!
